src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,136 @@
+        private static readonly LocalizableString MessageFormat = new LocalizableResourceString(nameof(FeaturesResources.Format_string_0_contains_invalid_placeholder_1), FeaturesResources.ResourceManager, typeof(FeaturesResources));
+        private static readonly LocalizableString Description = new LocalizableResourceString(nameof(FeaturesResources.Invalid_format_string), FeaturesResources.ResourceManager, typeof(FeaturesResources));
+
+        protected static DiagnosticDescriptor Rule = new DiagnosticDescriptor(
+            DiagnosticId,
+            Title,
+            MessageFormat,
+            DiagnosticCategory.Compiler,
+            DiagnosticSeverity.Warning,
+            isEnabledByDefault: true,
+            description: Description);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public DiagnosticAnalyzerCategory GetAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
+
+        public bool OpenFileOnly(Workspace workspace) => false;
+
+        protected static bool TryGetFormatMethod(SyntaxNodeAnalysisContext context, int numberOfArguments, SymbolInfo symbolInfo, out IMethodSymbol method)
+        {
+            method = null;
+
+            if (symbolInfo.Symbol == null)
+            {
+                return false;
+            }
+
+            if (!symbolInfo.Symbol.ContainingSymbol.Equals(context.SemanticModel.Compilation.GetTypeByMetadataName(typeof(string).FullName)))
+            {
+                return false;
+            }
+
+            if (symbolInfo.Symbol.Kind != SymbolKind.Method)
+            {
+                return false;
+            }
+
+            method = (IMethodSymbol)symbolInfo.Symbol;
+            if (method == null)
+            {
+                return false;
+            }
+
+            // If the chosen overload ends with a params object[] but there are the same number of
+            // arguments and parameters, then an actual object[] must be being passed instead of
+            // passing one argument at a time. This object[] may have been created somewhere we
+            // can't analyze, so never squiggle in this case.
+            if (method.Parameters[method.Parameters.Length - 1].IsParams && numberOfArguments == method.Parameters.Length)
+            {
+                return false;
+            }
+
+            return true;
+        }
+
+        private static string RemoveEscapedBrackets(string formatString)
+        {
+            var removeEscapedBracketsRegex = new Regex("{{|}}");
+            var modifiedFormatString = removeEscapedBracketsRegex.Replace(formatString, "  ");
+            return modifiedFormatString;
+        }
+
+        protected static void ValidateAndReportDiagnostic(SyntaxNodeAnalysisContext context, bool hasIFormatProvider, int numberOfArguments, string formatString, Location formatStringLocation)
```

Comment:
💭 You could use `int numberOfNonPlaceholderParameters` instead of `hasIFormatProvider` if you wanted to generalize this a bit more, though I cannot say definitively that it would be an improvement.